.gitpod.yml

```diff
@@ -6,16 +6,11 @@ ports:
     onOpen: open-preview
 tasks:
 - init: |
-    composer install --no-interaction --prefer-dist --no-scripts
-    composer clearcache
-    cp -R /home/gitpod/octobercms-config-docker config/docker
-    echo 'config/docker' >> .git/info/exclude
-    echo 'plugins/october/drivers' >> .git/info/exclude
     echo "APP_ENV=docker" > .env
-    echo "OCTOBERCMS_DEVELOP_COMMIT=$(git rev-parse HEAD)" >> .env
-    touch storage/database.sqlite
-    chmod 666 storage/database.sqlite
-    php artisan october:up
-    php artisan plugin:install october.drivers
+    echo "CMS_LINK_POLICY=secure" >> .env
+    cp -R /var/www/html/vendor .
```

Comment:
I understand that `cp -R /var/www/html/vendor .` is clever because it eliminates the need to run `composer install`.

BUT, it assumes that the `git clone` from your docker build packaged exactly the same `composer.json` as the one that's being pulled by Gitpod's git clone.

I fear that this assumption is likely violated when working with branches, particularly old maintenance branches. Gitpod getting one more commit than what the docker build got can be enough. 

As a result, the dependencies in `vendor` would be outdated/mismatching.

Alternative solution
* keep `composer install`
* to ensure that user's don't have to wait for `composer install`, configure the [Gitpod App](https://github.com/apps/gitpod-io/) on the repo. The app runs `tasks.init` from the `.gitpod.yml` on git-push.

Reply:
Agreed, the potential dependency mismatch is a concern. I assumed we'd setup a webhook to trigger the build, much like we're already doing to build the base image.

Using the Gitpod app may be the better alternative but does it build an image on the fly each time? I'm guessing that'll just force a developer to wait for the environment to be deployed to Gitpod instead of watching composer install in the foreground...

In any case, I'd advocate for a quicker deployment.

Reply:
1) The Gitpod App builds the *image* only when the Dockerfile changes. And if your `.gitpod.yml` refers to a docker image (and not a `Dockerfile`), then the App will never build the image, because it assumes it's already there. 

2) However, for every git-push the App does build a `snapshot.tar`. The `snapshot.tar` is created by running `git clone`, then all the commands from `tasks.init` from the `.gitpod.yml`. When finished the `snapshot.tar` is stored.... just like a stopped workspace. 

We call (1+2) *prebuilt workspaces*.

When you (or me) opens a new Gitpod workspace, Gitpod will check if there is already a prebuilt workspace for that specific commit. If found, it will just unpack the `snapshot.tar` into the `/workspace` folder into the new workspace container. `/workspace` is a mounted SSD-backed volume.

If there is no prebuilt workspace, for example, because the App is not installed, then Gitpod will run (1+2) while the user waits. 

If a prebuild has started, but not finished, when a user opens a workspace... the user can choose to wait for the prebuild (and follow the log output) or skip the prebuild and run everything like there was no prebuild. 

in practice, I rarely had to wait for a prebuild, because there are usually several minutes between one person pushing a commit or creating a PR and the next person wanting to open a workspaces on it. Enough minutes for the prebuild to finish.

Reply:
Cool. Your infrastructure sound super cool!

In practice, we've experienced issues with running composer install at runtime. Patch level updates pulled in from Packagist has occasionally broken things. By relying on the prebuilt image we can run tests and verify before it gets pushed.